pss/trojan_chunk.go

```diff
@@ -0,0 +1,214 @@
+	address chunk.Address
+	trojanData
+}
+
+// newTrojanChunk creates a new trojan chunk structure for the given address and message
+func newTrojanChunk(address chunk.Address, message trojanMessage) (*trojanChunk, error) {
+	chunk := &trojanChunk{
+		address: address,
+		trojanData: trojanData{
+			trojanHeaders: newTrojanHeaders(),
+			trojanMessage: message,
+		},
+	}
+	// find nonce for chunk
+	if err := chunk.setNonce(); err != nil {
+		return nil, err
+	}
+	return chunk, nil
+}
+
+// newTrojanHeaders creates an empty trojan headers struct
+func newTrojanHeaders() trojanHeaders {
+	// TODO: what should be the value of this?
+	span := make([]byte, 8)
+	// create initial nonce
+	nonce := make([]byte, 32)
+
+	return trojanHeaders{
+		span:  span,
+		nonce: nonce,
+	}
+}
+
+// setNonce determines the nonce so that when the trojan chunk fields are hashed, it falls in the neighbourhood of the trojan chunk address
+func (tc *trojanChunk) setNonce() error {
+	BMThashFunc := storage.MakeHashFunc(storage.BMTHash)() // init BMT hash function
+	err := iterateNonce(tc, BMThashFunc)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// iterateNonce iterates the BMT hash of the trojan chunk fields until the desired nonce is found
+func iterateNonce(tc *trojanChunk, hashFunc storage.SwarmHash) error {
+	// start out with random nonce
+	nonce := make([]byte, 32)
+	if _, err := rand.Read(nonce); err != nil {
+		return err
+	}
+	tc.nonce = nonce
+
+	// hash trojan chunk fields with different nonces until a desired one is found
+	hashWithinNeighbourhood := false // TODO: this could be correct on the 1st try
+	// TODO: add limit to tries
+	for hashWithinNeighbourhood != true {
+		serializedTrojanData, err := json.Marshal(tc.trojanData)
+		if err != nil {
+			return err
+		}
+		if _, err := hashFunc.Write(serializedTrojanData); err != nil {
+			return err
+		}
+		hash := hashFunc.Sum(nil)
+
+		// TODO: what is the correct way to check if hash is in the same neighbourhood as trojan chunk address?
+		_ = chunk.Proximity(tc.address, hash)
+
+		// TODO: replace placeholder condition
+		if true {
+			// if nonce found, stop loop
+			hashWithinNeighbourhood = true
+		} else {
+			// else, add 1 to nonce and try again
+			// TODO: find non sinful way of adding 1 to byte slice
+			// TODO: implement loop-around
+			nonceInt := new(big.Int).SetBytes(tc.nonce)
+			tc.nonce = nonceInt.Add(nonceInt, big.NewInt(1)).Bytes()
+		}
+	}
+
+	return nil
+}
+
+// toContentAddressedChunk creates a new addressed chunk structure with the given trojan message content serialized as its data
+func (tc *trojanChunk) toContentAddressedChunk() (chunk.Chunk, error) {
+	var emptyChunk = chunk.NewChunk([]byte{}, []byte{})
+
+	chunkData, err := json.Marshal(tc.trojanData)
+	if err != nil {
+		return emptyChunk, err
+	}
+	return chunk.NewChunk(tc.address, chunkData), nil
+}
+
+// equals compares the underlying data of 2 trojanData variables and returns true if they match, false otherwise
```

Comment:
what is this meant to be for?

Reply:
just a utility function used to check correctness when serializing/de-serializing trojan chunks. it can probably be moved to the test file

Reply:
moved to the test file, but there's still an unresolved related issue documented in the code (which is: why does a `reflect.DeepEquals` call does not work on 2 `trojanMessage`-typed vars?)